envoy/server.go

```diff
@@ -0,0 +1,437 @@
+	// TODO: Move this code in a common module and de-duplicate in sidecarhttp/envoy_api.go
+	// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+	// Used to join service name and port. Must not occur in service names
+	ServiceNameSeparator = ":"
+)
+
+type xdsCallbacks struct{}
+
+func (*xdsCallbacks) OnStreamOpen(context.Context, int64, string) error  { return nil }
+func (*xdsCallbacks) OnStreamClosed(int64)                               {}
+func (*xdsCallbacks) OnStreamRequest(int64, *api.DiscoveryRequest) error { return nil }
+func (*xdsCallbacks) OnStreamResponse(_ int64, req *api.DiscoveryRequest, _ *api.DiscoveryResponse) {
+	if req.GetErrorDetail().GetCode() != 0 {
+		log.Errorf("Received Envoy error code %d. Check Envoy logs for details.", req.GetErrorDetail().GetCode())
+		log.Debugf("Envoy error: %s", req.GetErrorDetail().GetMessage())
+	}
+}
+func (*xdsCallbacks) OnFetchRequest(context.Context, *api.DiscoveryRequest) error   { return nil }
+func (*xdsCallbacks) OnFetchResponse(*api.DiscoveryRequest, *api.DiscoveryResponse) {}
+
+// Server is a wrapper around Envoy's control plane xDS gRPC server and it uses
+// the Aggregated Discovery Service (ADS) mechanism.
+type Server struct {
+	Listener      *Listener
+	config        config.EnvoyConfig
+	state         *catalog.ServicesState
+	snapshotCache cache.SnapshotCache
+	xdsServer     xds.Server
+}
+
+// Run sets up the Sidecar listener event loop and starts the Envoy gRPC server
+func (s *Server) Run(ctx context.Context, looper director.Looper) {
+	grpcServer := grpc.NewServer()
+	discovery.RegisterAggregatedDiscoveryServiceServer(grpcServer, s.xdsServer)
+
+	go looper.Loop(func() error {
+		// Discard the event since we need a snapshot of the entire state
+		<-s.Listener.eventsChan
+
+		// When a server is expired in catalog/services_state.go -> ExpireServer(),
+		// the listener will receive an event for each expired service. We want to
+		// flush the channel to prevent rapid-fire updates to Envoy.
+		// This was inspired from receiver/receiver.go -> ProcessUpdates().
+		//
+		// TODO: We might want to consider flushing it in an infinite loop until
+		// it stays empty for a given time (say 500ms) because the channel can
+		// buffer at most catalog.LISTENER_EVENT_BUFFER_SIZE
+		pendingEventCount := len(s.Listener.eventsChan)
+		for i := 0; i < pendingEventCount; i++ {
+			<-s.Listener.eventsChan
+		}
+
+		listeners, err := s.EnvoyListenersFromState()
+		if err != nil {
+			log.Errorf("Failed to create Envoy listeners: %s", err)
+			return nil
+		}
+
+		version := time.Now().Unix()
+		err = s.snapshotCache.SetSnapshot(
+			// This needs to match the value passed via `--service-node` to Envoy
+			// See https://github.com/envoyproxy/envoy/issues/144#issuecomment-267401271
+			s.state.Hostname,
+			cache.NewSnapshot(
+				strconv.FormatInt(version, 10),
+				nil,
+				s.EnvoyClustersFromState(),
```

Comment:
Yeah, seems like we should just make an `envoy_adapter` package to hold this stuff.

Reply:
I ended up creating package `envoy/adapter`, to avoid breaking the go style guide of not using underscores in package names.